File: LeetCode Solutions/maximalRectangle.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

class Solution {
    public:
        int maximalRectangle(vector<vector<char>>& matrix){
            if(matrix.empty())
                return 0;

            int ans = 0;
            vector<int> hist(matrix[0].size());

            //Iterate through each row in the matrix
            for (const vector<char>& row : matrix){
                //update histogram for each row
                for (int i = 0; i < row.size(); ++i)
                {
                    hist[i] = row[i] == '0' ? 0 : hist[i] + 1;
                }
                //calculate the largest rectangle area using the histogram of the current row.
                ans = max(ans, largestRectangleArea(hist));
            }

            return ans;
        }
    private:
        int largestRectangleArea(const vector<int>& heights){
            int ans = 0;
            stack<int> stack;

            //interate through each element in the heights array and calculate the largest area
            for (int i = 0; i <= heights.size(); ++i)
            {
                while (!stack.empty() && (i == heights.size() || heights[stack.top()] > heights[i]))
                {
                    //calculate the area of the rectangle with the current height
                    const int h = heights[stack.top()];
                    stack.pop();
                    const int w = stack.empty() ? i : i - stack.top() - 1;
                    ans = max(ans, h*w);
                }
                stack.push(i);
            }
            
            return ans;
        }
};

int main() {
    // Example binary matrix
    vector<vector<char>> matrix = {
        {'1', '0', '1', '0', '0'},
        {'1', '0', '1', '1', '1'},
        {'1', '1', '1', '1', '1'},
        {'1', '0', '0', '1', '0'}
    };

    Solution sol; // Create an instance of the Solution class

    // Call the maximalRectangle method and print the result
    cout << "Maximal Rectangle Area: " << sol.maximalRectangle(matrix) << endl;

    return 0;
}

File: LeetCode Solutions/83. Maximal Rectangle/maximalRectangle.py
class Solution:
    def maximalRectangle(self, matrix):
        if not matrix:
            return 0

        ans = 0
        hist = [0] * len(matrix[0])

        for row in matrix:
            for i in range(len(row)):
                hist[i] = 0 if row[i] == '0' else hist[i] + 1
            ans = max(ans, self.largestRectangleArea(hist))

        return ans

    def largestRectangleArea(self, heights):
        ans = 0
        stack = []

        for i in range(len(heights) + 1):
            while stack and (i == len(heights) or heights[stack[-1]] > heights[i]):
                h = heights[stack.pop()]
                w = i if not stack else i - stack[-1] - 1
                ans = max(ans, h * w)
            stack.append(i)

        return ans

# Example usage
if __name__ == "__main__":
    # Example binary matrix
    matrix = [
        ['1', '0', '1', '0', '0'],
        ['1', '0', '1', '1', '1'],
        ['1', '1', '1', '1', '1'],
        ['1', '0', '0', '1', '0']
    ]

    sol = Solution()
    print("Maximal Rectangle Area:", sol.maximalRectangle(matrix))